contracts/contract_call.go

```diff
@@ -0,0 +1,162 @@
+package contracts
+
+import (
+	"math/big"
+
+	"github.com/celo-org/celo-blockchain/accounts/abi"
+	"github.com/celo-org/celo-blockchain/common"
+	"github.com/celo-org/celo-blockchain/common/hexutil"
+	"github.com/celo-org/celo-blockchain/core/vm"
+	"github.com/celo-org/celo-blockchain/log"
+)
+
+// Call represent a runnable call on the EVM
+type Call interface {
+	// Run will execute the call using the caller.
+	// Execution result will be populated into result
+	Run(caller vm.EVMCaller, result interface{}) (leftoverGas uint64, err error)
```

Comment:
I'm not convinced about this design. Where previously we would call a method on a VM instance we now:

Construct Call object -> Call Run on call object with EVMCaller as param -> Run then calls Method on EVMCaller -> EVMCaller then calls method on VM instance.

This seems like a lot of redirection, and to me makes understanding the path of calls to the VM more difficult to follow.

The original problem that I think we were trying to solve here was that we were using VM instances more than once, looking at the history of this #938 it looks like people were well aware of the fact that they were re-using VM instances, Its not clear whether they realised they should not be doing this though.

The approach taken here is to use the EVMCaller construct which should then be used everywhere and always ensures a new instance of the VM is used for each call. This then allows us to have calls that internally make multiple calls.

I feel like this could also be solved by just making it easy to create a new VM instance and ensuring that we use a new instance per call. We could do this by funnelling our calls to the VM through a higher level function that ensures that we use a new VM instance per call (for example something like the Run method here that takes a vm factory in place of the EVMCaller) . This would reduce the hierarchy of calls to get to the evm (just, we call Run -> Run calls VM).

I prefer the approach of making VM instantiation easy, as opposed to the EVMCaller because I think, beyond reducing the call heirarchy depth to get to the VM, a factory is a more flexible component that is less likely to need to be changed as the code changes.

Both a factory and EVMCaller will need to be changed when the required parametrs to construct a VM change (hopefully this should not be too often, but we don't really have any control of this because of upstream).

Where the two designs differ is that if we ever need access to any of the 12 other public methods of the vm we need to then either expose them through the EVMCaller interface or at that point create a factory to give ourselves a vm instance. Its not even clear to me how you would correctly expose those methods on an instance of the EVMCaller, for example, for the  the `Cancelled` method we would then need to store old evm instances in the EVMCaller to be able to retrieve the result of `Cancelled` but if the EVMCaller is allowing multiple internal EVMs to be constructed in its lifetime how do we expose `Cancelled` correctly?

Right now given that it looks like our contract calling code is going to change considerably I think providing a factory and handling evm instances on a case by case basis will probably introduce less friction. Once we have clearer idea of how this code is going to end up we may then see that an EVMCaller is a useful component.

I've made these changes in another branch, see [here](https://github.com/celo-org/celo-blockchain/compare/mc/evm-caller..731425bc2f935729d1bf2d0ce0f88a4b053d1292#diff-f64965fa055835e4285f91971de4d4aec7c1a4a4c47aedb8e4c58761517115cdR69-R79) for how calling methods have changed.  Essentially there is no EVMCaller any more just a factory and provider, and we don't have the concept of "multi call" calls instead callers pass a new evm instance to each `Call.Run` invocation and `Call.Run` invocations only make one call. These changes resulted in the dropping of about 150 lines of code which I think is a significant enough change to point to the fact that this is a simplification.

Reply:
I don't agree :) 

To me, sending an evm vs sending an evm caller implies that:

* **sending evm:** we rely on a social contract for developers not to missuse the evm to make contract calls
* **sending evm caller**: we rely on code to ensure the contract

I will always favor code. I don't think it adds a ton of complexity, nor it makes it unreadable. 

I'm happy to discuss the other aspects, maybe having a `Call` interface doesn't make sense. Maybe having 2 types of calls doesn't make sense... i'm not bullish about that, i was trying to make calls more "clear" (vs sending 9 positional arguments to a function); but surely other was can be found to do the same thing.

About the topic of "accessing other evm methods". Please check the `vm/runtime` package. It states 
```
// Package runtime provides a basic execution model for executing EVM code.
```

It exposes `runtime.NewEnv() runtinme.Config` and then methods `Call`, `Create`, `Execute` that receve the config and will create an EVM, do a single call, and then discard that.

What i'm doing is basically that. Maybe `EVMCaller` is a bad name, since gives you the sense that let's you access to the full EVM, and maybe instead should be called `ContractCaller` and hide the fact that an EVM is inside. To a contract you either want to do queries or exeuctions. The rest is not important.

Reply:
>To me, sending an evm vs sending an evm caller implies that:
>
>sending evm: we rely on a social contract for developers not to missuse the evm to make contract calls
>sending evm caller: we rely on code to ensure the contract

Maybe I didn't explain myself very well, I too generally favour code to enforce things.

What I was suggesting is that we could enforce the use of the EVM at a different level.

So currently in this PR the EVM use is enforced at the following level.

BoundMethod.Execute -> evmCaller.Execute (vm non reuse is enforced here)

I was suggesting that we could instead enforce this in BoundMethod.Execute, by having it and its sibling methods accept a VM factory or provider in the place of the caller.

I don't think this introduces a reliance on a social contract, in both cases users will call BoundMethod.XXX and the contract call package handles enforcing correct vm usage. In either case there is still a reliance on social contract, which is that users call BoundMethod.XXX, if you are not calling BoundMethod.XXX then you can do what you like, neither approach enforces anything at that point.

The contract_comm api would hardly change, what I'm suggesting is
```go
	m.VMExecute(provider, returnObj, value, args...)
	// as opposed to 
	m.VMExecute(caller, returnObj, value, args...)
```

I'm not clear about what your point is with the runtime package. Maybe I'm missing some context but I just wonder if we can be sure that we wont need to access other VM methods?

The question of needing other methods aside, If both implementations are providing the same level of enforcement over the correct use of the VM why not choose the solution that has 150 less LOC?

Reply:
To be honest @piersy . I don't care as much. The way is see it, i've spent several days going over and over the whole codebase, trying to understand it all, and find a way to simplify the way we do call to contracts, and then migrate all the calls to the system.

Where it's better to call it provider or caller, makes no difference (but implies changing it again). I don't buy the LOC argument. A singleton surely takes a lot less LOC than passing a service all over the system to achieve the same goal. Does that make the singleton simpler or easier to understand? I don't think so. It's subjective, there are times where LOC can be a clearer indication on complexity, others where they won't be. At least that my opinion. There are several ways to measure complexity of code. So i don't appreciate being measured by just a single one.

I see this discussion as going to an annoying level of detail about a minor thing, when the much bigger thing (cleaning this up, and migrating the contracts) it's not being addressed.

If you strongly believe that provider is strictly better, then go ahead and replace everything with Provider (And please change https://github.com/celo-org/celo-blockchain/pull/1522 in the process). Just be sure that the notion of no reuse of evm doesn't leave the BoundContract methods...  Also, have in mind that `EVMProvider` implies it creates a EVM.. but what kind of EVM? In particular we create a specific evmContext `context := New(common.ZeroAddress, common.Big0, header, chain, nil)` which only makes sense for this type of calls; so i would include that in the name. `EVMProvider` seems too generic.

Reply:
Hey @mcortesi I do think the provider is a slightly simpler approach that would provide more flexibility in future changes, but at this point I'm not sure its worth switching and changing everything in #1522. If we run into problems with the EVMCaller we can always re-consider the provider route.

Good point regarding the singleton, I can see how you could drop some LOC with one and it's almost never going to be an improvement.

I think at a certain point though everyone has to buy the LOC argument, anyone would be hard pressed to argue that 10,000 LOC is simpler than 1000 LOC that provides the same functionality. The question is about where you draw your thresholds.

In this case an extra layer was added the `EVMCaller` and that added LOC, and in my opinion complexity because it adds an extra layer of indirection to get to the EVM, but it also adds the feature of ensuring non reuse of the EVM, and you can't have features without LOC :)

Reply:
Totally, at some point a big difference in LOC speaks for itself.

Let's keep it as it is, and keep an eye on the future to see if needs further changes or to move to provide. In a way, this is a refactor, but not necessarily the last one; we can always refine later. What is important, is to be sure it's in the right direction (aka. "better than before")